File: src/2024/day9.zig
const std = @import("std");
const utils = @import("utils");

pub fn main() !void {
    // https://adventofcode.com/2024/day/9
    utils.printTitle(2024, 9, "Disk Fragmenter");

    const m = utils.Monitor.init();
    defer m.deinit();
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const ally = gpa.allocator();
    const input = comptime utils.trimInput(@embedFile("input"));

    const p1 = try part1(input, ally);
    utils.printAnswer(1, p1);
    std.debug.assert(6288707484810 == p1);

    const p2 = try part2(input, ally);
    utils.printAnswer(2, p2);
    std.debug.assert(6311837662089 == p2);
}

const Block = struct {
    start: usize,
    len: usize,
};

const Disk = struct {
    files: std.ArrayList(Block),
    freespaces: std.ArrayList(Block),
    data: std.ArrayList(?u32),
    nextFileId: u32,
    ally: std.mem.Allocator,

    pub fn parse(input: []const u8, ally: std.mem.Allocator) !Disk {
        var files = std.ArrayList(Block).empty;
        var freespaces = std.ArrayList(Block).empty;
        var disk = std.ArrayList(?u32).empty;
        var nextId: u32 = 0;

        for (0.., input) |i, c| {
            const blockLen = try std.fmt.charToDigit(c, 10);
            if (i & 1 == 0) {
                std.debug.assert(blockLen != 0);
                try files.append(ally, .{ .start = disk.items.len, .len = blockLen });
                try disk.appendNTimes(ally, nextId, blockLen);
                nextId += 1;
            } else if (blockLen != 0) {
                try freespaces.append(ally, .{ .start = disk.items.len, .len = blockLen });
                try disk.appendNTimes(ally, null, blockLen);
            }
        }

        return .{
            .files = files,
            .freespaces = freespaces,
            .data = disk,
            .nextFileId = nextId,
            .ally = ally,
        };
    }

    pub fn deinit(self: *@This()) void {
        self.files.deinit(self.ally);
        self.freespaces.deinit(self.ally);
        self.data.deinit(self.ally);
    }

    pub fn checksum(self: *const @This()) u64 {
        var result: u64 = 0;
        for (0.., self.data.items) |i, n| {
            if (n) |num| {
                result += i * num;
            }
        }

        return result;
    }
};

fn part1(input: []const u8, ally: std.mem.Allocator) !u64 {
    var disk = try Disk.parse(input, ally);
    defer disk.deinit();

    var front: usize = 0;
    var back = disk.data.items.len;
    while (back > front) {
        back -= 1;

        if (disk.data.items[back] == null) {
            continue;
        }

        while (front < disk.data.items.len and disk.data.items[front] != null) {
            front += 1;
        }

        if (back > front) {
            std.mem.swap(?u32, &disk.data.items[front], &disk.data.items[back]);
            front += 1;
        }
    }

    return disk.checksum();
}

test "part 1" {
    const example =
        \\2333133121414131402
    ;
    try std.testing.expectEqual(1928, try part1(example, std.testing.allocator));
}

fn part2(input: []const u8, ally: std.mem.Allocator) !u64 {
    var disk = try Disk.parse(input, ally);
    defer disk.deinit();
    var nextId = disk.nextFileId;

    while (nextId != 0) {
        nextId -= 1;
        const fileBlock = disk.files.items[nextId];

        for (disk.freespaces.items) |*freeBlock| {
            if (freeBlock.start >= fileBlock.start) break;
            if (freeBlock.len >= fileBlock.len) {
                var f = freeBlock.start;
                var r = fileBlock.start;
                for (0..fileBlock.len) |_| {
                    std.mem.swap(?u32, &disk.data.items[f], &disk.data.items[r]);
                    f += 1;
                    r += 1;
                }
                freeBlock.len -= fileBlock.len;
                freeBlock.start += fileBlock.len;
                break;
            }
        }
    }

    return disk.checksum();
}

test "part 2" {
    const example =
        \\2333133121414131402
    ;
    try std.testing.expectEqual(2858, try part2(example, std.testing.allocator));
}

File: src/2024/day9.rs
#[path = "../utils.rs"]
mod utils;

fn main() {
    // https://adventofcode.com/2024/day/9
    println!("Day 9, 2024: Disk Fragmenter");
    let _m = utils::Monitor::start();

    let p1 = part1(get_input());
    println!("  Part 1: {}", p1);
    assert_eq!(6288707484810, p1);

    let p2 = part2(get_input());
    println!("  Part 2: {}", p2);
    assert_eq!(6311837662089, p2);
}

fn part1(input: &'static str) -> u64 {
    let mut disk = Disk::parse(input);
    disk.defrangment();
    disk.checksum()
}

fn part2(input: &'static str) -> u64 {
    let mut disk = Disk::parse(input);
    disk.defragment_whole_files();
    disk.checksum()
}

struct Block {
    start: usize,
    len: usize,
}

struct Disk {
    data: Vec<Option<u64>>,
    files: Vec<Block>,
    free_spaces: Vec<Block>,
    next_file_id: u64,
}

impl Disk {
    pub fn checksum(self: &Self) -> u64 {
        self.data
            .iter()
            .enumerate()
            .fold(0, |acc, (i, c)| acc + (i as u64) * c.unwrap_or(0))
    }

    pub fn parse(input: &'static str) -> Self {
        let mut data = Vec::<Option<u64>>::new();
        let mut files = Vec::<Block>::new();
        let mut free_spaces = Vec::<Block>::new();
        let mut next_file_id = 0;

        for (i, c) in input.bytes().enumerate() {
            let len = (c - b'0') as usize;
            let block = Block {
                start: data.len(),
                len: len,
            };

            if i & 1 == 0 {
                files.push(block);
                for _ in 0..len {
                    data.push(Some(next_file_id));
                }
                next_file_id += 1
            } else {
                free_spaces.push(block);
                for _ in 0..len {
                    data.push(None);
                }
            }
        }

        Self {
            data: data,
            files: files,
            free_spaces: free_spaces,
            next_file_id: next_file_id,
        }
    }

    pub fn defrangment(self: &mut Self) {
        let mut back = self.data.len();
        let mut front = 0;
        let len = self.data.len();

        while back > front {
            back -= 1;

            if let None = self.data[back] {
                continue;
            }

            while front < len {
                if let Some(_) = self.data[front] {
                    front += 1;
                } else {
                    break;
                }
            }

            if back > front {
                self.data[front] = self.data[back];
                self.data[back] = None;
                front += 1
            }
        }
    }

    pub fn defragment_whole_files(self: &mut Self) {
        let mut last_file = self.next_file_id;

        while last_file != 0 {
            last_file -= 1;
            let file_block = &self.files[last_file as usize];

            for free_block in &mut self.free_spaces {
                if free_block.start >= file_block.start {
                    break;
                }

                if free_block.len >= file_block.len {
                    let mut front = free_block.start;
                    let mut back = file_block.start;

                    for _ in 0..file_block.len {
                        self.data[front] = self.data[back];
                        self.data[back] = None;
                        front += 1;
                        back += 1;
                    }

                    free_block.len -= file_block.len;
                    free_block.start += file_block.len;
                    break;
                }
            }
        }
    }
}

fn get_input() -> &'static str {
    include_str!("../../inputs/2024/day9.txt").trim_end()
}

#[cfg(test)]
mod tests {
    fn get_example() -> &'static str {
        "2333133121414131402"
    }

    #[test]
    fn part1() {
        assert_eq!(1928, super::part1(get_example()));
    }

    #[test]
    fn part2() {
        assert_eq!(2858, super::part2(get_example()));
    }
}
